src/pkgcheck/checks/reserved.py

```diff
@@ -0,0 +1,94 @@
+import re
+from .. import addons, bash, results, sources
+from . import Check
+
+
+class _ReservedNameCheck(Check):
+    reserved_prefixes = ('__', 'abort', 'dyn', 'prep')
+    reserved_substrings = ('hook', 'paludis', 'portage')  # 'ebuild' is special case
+    reserved_ebuild_regex = re.compile(r'(.*[^a-zA-Z])?ebuild.*')
+
+    """Portage variables whose use is half-legitimate and harmless if the package manager doesn't support them"""
+    special_whitelist = ('EBUILD_DEATH_HOOKS', 'EBUILD_SUCCESS_HOOKS', 'PORTAGE_QUIET')
+
+    def _feed(self, item):
+        function_names = {
+            item.node_str(node.child_by_field_name('name')): node.start_point
+            for node, _ in bash.func_query.captures(item.tree.root_node)
+        }
+        variable_names = {
+            item.node_str(node.child_by_field_name('name')): node.start_point
+            for node, _ in bash.var_assign_query.captures(item.tree.root_node)
+        }
+
+        for used_type, used_names in (('function', function_names), ('variable', variable_names)):
+            for used_name, node_start in used_names.items():
```

Comment:
This is a weird loop construction to me. I would maybe create another method that holds the inner body and have two independent loops:

```python
def _check(self, item, used_type, nodes):
    for node in nodes:
        name = item.node_str(node.child_by_field_name('name')
        if name in self.special_whitelist:
            continue
        test_name = name.lower()
        lineno, _ = node.start_point
        for reserved in self.reserved_prefixes:
            if test_name.startswith(reserved):
                yield name, used_type, reserved, 'prefix', lineno
        for reserved in self.reserved_substrings:
            if reserved in test_name:
                yield name, used_type, reserved, 'substring', lineno
        if self.reserved_ebuild_regex.match(test_name):
             yield used_name, used_type, 'ebuild', 'substring', lineno

def _feed(self, item):
    yield from self._check(item, 'function', node for node, _ in bash.func_query.captures(item.tree.root_node))
    yield from self._check(item, 'variable', node for node, _ in bash.var_assign_query(item.tree.root_node))
```

but I get that this is a matter of taste, so it's just a suggestion from my side :-)

Reply:
Thanks for the suggestion. I still need to separate the extraction of name from node, as I need the `dict` so it shows the warning only once per name. But it indeed looks better now - thank you.

Reply:
Ah, I didn't realise that dict was also to have some deduplication, that makes sense. Just out of curiosity, which occurrence will then be reported? The last I take it?